quil-rs/src/expression/simplification/mod.rs

```diff
@@ -0,0 +1,153 @@
+use crate::expression::Expression;
+
+mod by_hand;
+
+/// Simplify an [`Expression`].
+pub(super) fn run(expression: &Expression) -> Expression {
+    by_hand::run(expression)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::str::FromStr;
+
+    macro_rules! test_simplify {
+        ($name:ident, $input:expr, $expected:expr) => {
+            #[test]
+            fn $name() {
+                let parsed_input = Expression::from_str($input)
+                    .unwrap_or_else(|error| panic!("Parsing input `{}` failed: {error}", $input));
+                let parsed_expected = Expression::from_str($expected).unwrap_or_else(|error| {
+                    panic!(
+                        "Parsing expected expression `{}` failed: {error}",
+                        $expected
+                    )
+                });
+                let computed = run(&parsed_input);
+                assert_eq!(parsed_expected, computed);
+            }
+        };
+    }
+
```

Comment:
I think my only remaining question is about test coverage.  It's not obvious to me how many of the cases in `by_hand` are getting exercised by these tests.

Reply:
Solid point. All the test cases are ported from the current unit testing, plus a few more for specific rules, but I should put in more tests.